### ISSUE TITLE
Local invoke fails when trying to use Layers locally

### ISSUE THREAD

**Issue by heitorlessa** [awslabs/aws-sam-cli#816]
**Description:**

When trying to use the new Layers feature locally it fails due to not finding a Dockerfile which seems to be created at runtime -- Haven't dig too much into it yet as I just bumped into that.

**Steps to reproduce the issue:**
1. **Initialize a boilerplate app**: `sam init -r python; cd sam-app`
2. **Install dependencies into a lib folder**: `pip install -r hello_world/requirements.txt -t lib/`
3. **Create a Layer** in `template.yaml`:

```yaml
    RequestsLayer:
        Type: 'AWS::Serverless::LayerVersion'
        Properties:
            ContentUri: lib/
```

4. **Reference that Layer** in the `HelloWorldFunction` resource:

```yaml
    HelloWorldFunction:
        Type: AWS::Serverless::Function
        Properties:
            CodeUri: hello_world/
            Handler: app.lambda_handler
            Runtime: python3.6 # also tried python3.7 (default)
            Layers:
                - !Ref RequestsLayer
```

5. Run function locally: ``sam local invoke HelloWorldFunction --no-event``

**Observed result:**

Local invocation fails quickly and returning the trace data at the bottom.

**Expected result:**

Build Layer locally and local invocation to succeed.

**Additional environment details (Ex: Windows, Mac, Amazon Linux etc)** OSX Sierra 10.12.6 (16G1618)

**Output of `sam --version`**: SAM CLI, version 0.8.0

**Optional Debug logs**:

`Add --debug flag to command you are running`

```python
sam local invoke HelloWorldFunction --no-event
2018-11-30 08:42:46 Found credentials in shared credentials file: ~/.aws/credentials
2018-11-30 08:42:46 Invoking app.lambda_handler (python3.6)
2018-11-30 08:42:46 RequestsLayer is a local Layer in the template
2018-11-30 08:42:46 Image was not found.
2018-11-30 08:42:46 Building image...
Traceback (most recent call last):
  File "/Users/lessa/.pyenv/versions/work/bin/sam", line 11, in <module>
    sys.exit(cli())
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/commands/local/invoke/cli.py", line 56, in cli
    parameter_overrides)  # pragma: no cover
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/commands/local/invoke/cli.py", line 99, in do_cli
    stderr=context.stderr)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/commands/local/lib/local_lambda.py", line 75, in invoke
    self.local_runtime.invoke(config, event, debug_context=self.debug_context, stdout=stdout, stderr=stderr)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/local/lambdafn/runtime.py", line 81, in invoke
    debug_options=debug_context)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/local/docker/lambda_container.py", line 66, in __init__
    image = LambdaContainer._get_image(image_builder, runtime, layers)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/local/docker/lambda_container.py", line 157, in _get_image
    return image_builder.build(runtime, layers)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/local/docker/lambda_image.py", line 113, in build
    self._build_image(base_image, image_tag, downloaded_layers)
  File "/Users/lessa/.pyenv/versions/3.6.2/envs/work/lib/python3.6/site-packages/samcli/local/docker/lambda_image.py", line 173, in _build_image
    with open(str(full_dockerfile_path), "w") as dockerfile:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/lessa/.aws-sam/layers-pkg/dockerfile_3a7f092d-0b92-42c9-b823-19c2401f44c5'
```

**Comment by timoschilling** [comment]
I can confirm this bug using `Runtime: ruby2.5`.

I even have a problem packaging this setup. The packaging didn't uploads a layer to S3 and didn't changes the `ContentUri` property.
```yaml
AWSTemplateFormatVersion: 2010-09-09
Parameters:
Resources:
  Lambda:
    Properties:
      CodeUri: s3://sam-applications/12ed7a722e38966db52f8b5d9bd0687d
      Events:
        Callback:
          Properties:
            Method: post
            Path: /
          Type: Api
      Handler: app.lambda_handler
      Layers:
      - Ref: LibLayer
      Runtime: ruby2.5
    Type: AWS::Serverless::Function
  LibLayer:
    Properties:
      ContentUri: lib/
    Type: AWS::Serverless::LayerVersion
Transform: AWS::Serverless-2016-10-31
```

**Comment by PrivoAdam** [comment]
Adding a bit more to this - I ran into the same issue and manually created the `~/.aws-sam/layers-pkg` directory.   The next build fails with 

```
16:55 $ sam local invoke -e test-event.json test
2018-12-02 16:57:44 Found credentials in shared credentials file: ~/.aws/credentials
2018-12-02 16:57:44 Invoking test.handler (nodejs8.10)
2018-12-02 16:57:44 layerCore is a local Layer in the template
2018-12-02 16:57:44 Image was not found.
2018-12-02 16:57:44 Building image...
2018-12-02 16:57:45 Failed to build Docker Image
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/samcli/local/docker/lambda_image.py", line 187, in _build_image
    pull=not self.skip_pull_image)
  File "/usr/lib/python2.7/site-packages/docker/models/images.py", line 283, in build
    raise BuildError(chunk['error'], result_stream)
BuildError: Unknown flag: chown
Error: Building Image failed.
```

I did grab the dockerfile generated by sam prior to the process cleaning up:

```
FROM lambci/lambda:nodejs8.10
ADD --chown=sbx_user1051:495 layerCore /opt
```

**Comment by jfuss** [comment]
Please DO NOT manually create `~/.aws-sam/layers-pkg` for any reason! This directory is the cache for layers and other directories SAM CLI could create in the future. SAM CLI is downloading code to that directory and you are running the risk of having the directory improperly scoped.

We only create that directory on the [downloading of a layer](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/layers/layer_downloader.py#L84). We need to do this if you are referencing a layer. This bug will only impact you, if you are first trying to use layers without downloading a layer. 

@timoschilling What version of AWS CLI are you running? Make sure you have upgraded to the latest version otherwise you will not be able to package LayerVersions.

@PrivoAdam What version of docker are you running?

**Comment by PrivoAdam** [comment]
@jfuss Thanks for the clarification on the `~/.aws-sam/layers-pkg` directory.  

I'm running Fedora 28 with the default docker version 1.13.1.  I've moved to the docker CE repo now with 18.09.0 which supports --chown in ADD.  It looks like anyone using under docker 17.09.0 may run into this.

**Comment by timoschilling** [comment]
@jfuss I'm on `0.8.1`

**Comment by rodrigomaia** [comment]
I'm have the same bug of @timoschilling.

My version is 0.8.0

**Comment by jfuss** [comment]
@timoschilling That is SAM CLI's version not AWS CLI. SAM CLI shells out to AWS CLI for package and deploy. Therefore you need to update AWS CLI to a min version of 1.16.67. If there is some other issue that does not relate to the layer_cache being corrected, please cut a new issue so we can keep this issue focused on the original bug.

@rodrigomaia: Update AWS CLI to min version of 1.16.67.

**Comment by timoschilling** [comment]
@jfuss ok, works now, thanks. Why SAM CLI didn't requires an actual version of the AWS CLI?

**Comment by rodrigomaia** [comment]
Works... tks.

**Comment by H6** [comment]
@jfuss Hi i still have the same issue with `sam local invoke` and local layers. With python3.6 and the most current CLIs

`SAM CLI, version 0.8.1`
`aws-cli/1.16.74 Python/3.6.5 Darwin/18.2.0 botocore/1.12.64`
`Docker version 18.09.0, build 4d60db4`

```
Traceback (most recent call last):
  File "/Users/db/Library/Python/3.6/bin/sam", line 11, in <module>
    sys.exit(cli())
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/commands/local/invoke/cli.py", line 56, in cli
    parameter_overrides)  # pragma: no cover
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/commands/local/invoke/cli.py", line 99, in do_cli
    stderr=context.stderr)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/commands/local/lib/local_lambda.py", line 75, in invoke
    self.local_runtime.invoke(config, event, debug_context=self.debug_context, stdout=stdout, stderr=stderr)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/local/lambdafn/runtime.py", line 81, in invoke
    debug_options=debug_context)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/local/docker/lambda_container.py", line 66, in __init__
    image = LambdaContainer._get_image(image_builder, runtime, layers)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/local/docker/lambda_container.py", line 157, in _get_image
    return image_builder.build(runtime, layers)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/local/docker/lambda_image.py", line 113, in build
    self._build_image(base_image, image_tag, downloaded_layers)
  File "/Users/db/Library/Python/3.6/lib/python/site-packages/samcli/local/docker/lambda_image.py", line 173, in _build_image
    with open(str(full_dockerfile_path), "w") as dockerfile:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/db/.aws-sam/layers-pkg/dockerfile_9c562307-7ace-47e6-977f-9f285dbb6d8c'
```

**Comment by jfuss** [comment]
This was released in [v0.9.0](https://github.com/awslabs/aws-sam-cli/releases/tag/v0.9.0)

Closing

**Comment by gham-khaled** [comment]
I had the same Error: Invalid value for '--layer-cache-basedir' 
Solved it with changing the permissions of my /.aws-sam/layers-pkg directory
